src/primitives/hrpstring.rs

```diff
@@ -0,0 +1,558 @@
+    /// String does not contain the separator character.
+    #[non_exhaustive]
+    MissingSeparator,
+    /// The checksum does not match the rest of the data.
+    #[non_exhaustive]
+    InvalidChecksum,
+    /// The data payload is empty.
+    #[non_exhaustive]
+    InvalidDataEmpty,
+    /// The checksum is not a valid length.
+    #[non_exhaustive]
+    InvalidChecksumLength,
+    /// Attempt conversion of an invalid witness version string/number.
+    InvalidWitnessVersion(witness_version::ConversionError),
+    /// Some part of the string contains an invalid character.
+    InvalidChar(char),
+    /// The bit conversion failed due to a padding issue.
+    #[non_exhaustive]
+    InvalidPadding,
+    /// The whole string must be of one case.
+    #[non_exhaustive]
+    MixedCase,
+}
+
+impl From<HrpError> for Error {
+    fn from(e: HrpError) -> Self { Error::InvalidHrp(e) }
+}
+
+impl From<Infallible> for Error {
+    fn from(v: Infallible) -> Self { match v {} }
+}
+
+impl From<witness_version::ConversionError> for Error {
+    fn from(e: witness_version::ConversionError) -> Self { Error::InvalidWitnessVersion(e) }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use Error::*;
+
+        match *self {
+            InvalidHrp(ref e) => write_err!(f, "invalid human-readable part"; e),
+            MissingSeparator => write!(f, "missing human-readable separator, \"{}\"", SEP),
+            InvalidChecksum => write!(f, "invalid checksum"),
+            InvalidDataEmpty => write!(f, "invalid data - payload is empty"),
+            InvalidChecksumLength => write!(f, "the checksum is not a valid length"),
+            InvalidChar(n) => write!(f, "invalid character (code={})", n),
+            InvalidPadding => write!(f, "invalid padding"),
+            InvalidWitnessVersion(ref e) => write_err!(f, "witness version error"; e),
+            MixedCase => write!(f, "mixed-case strings not allowed"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use Error::*;
+
+        match *self {
+            InvalidWitnessVersion(ref e) => Some(e),
+            InvalidHrp(ref e) => Some(e),
+            MissingSeparator
+            | MixedCase
+            | InvalidChecksum
+            | InvalidDataEmpty
+            | InvalidChecksumLength
+            | InvalidChar(_)
+            | InvalidPadding => None,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::convert::TryFrom;
+
+    use super::*;
+
+    #[test]
+    fn try_from_err() {
+        assert!(Fe32::try_from(32_u8).is_err());
+        assert!(Fe32::try_from(32_u16).is_err());
+        assert!(Fe32::try_from(32_u32).is_err());
+        assert!(Fe32::try_from(32_u64).is_err());
+        assert!(Fe32::try_from(32_u128).is_err());
+    }
+
+    #[test]
+    #[allow(unused_variables)] // Triggered by matches macro.
```

Comment:
This is triggered because the `matches` macro is used incorrectly.

Reply:
Removed, and fixed tests to fully implement the test vectors from both bips.